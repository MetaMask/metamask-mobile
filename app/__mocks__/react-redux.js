export * from 'react-redux-test';
