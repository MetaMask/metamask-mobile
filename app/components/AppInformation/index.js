import React, { Component } from 'react';
import {
	SafeAreaView,
	StyleSheet,
	Image,
	Text,
	InteractionManager,
	View,
	ScrollView,
	TouchableOpacity,
	Linking
} from 'react-native';
import { colors, fontStyles } from '../../styles/common';
import PropTypes from 'prop-types';
import { strings } from '../../../locales/i18n';

const styles = StyleSheet.create({
	wrapper: {
		backgroundColor: colors.white,
		flex: 1
	},
	wrapperContent: {
		paddingLeft: 20,
		paddingRight: 20,
		paddingVertical: 20
	},
	title: {
		fontSize: 18,
		textAlign: 'left',
		marginBottom: 20,
		...fontStyles.normal
	},
	link: {
		fontSize: 18,
		textAlign: 'left',
		marginBottom: 20,
		...fontStyles.normal,
		color: colors.primary
	},
	division: {
		borderBottomColor: colors.gray,
		borderBottomWidth: 1,
		width: '30%',
		marginBottom: 20
	},
	image: {
		width: 100,
		height: 100
	},
	logoWrapper: {
		flex: 1,
		backgroundColor: colors.white,
		alignItems: 'center',
		justifyContent: 'center',
		top: 20,
		marginBottom: 40
	}
});

const foxImage = require('../../images/fox.png'); // eslint-disable-line import/no-commonjs

/**
 * View that contains app information
 */
export default class AppInformation extends Component {
	static navigationOptions = () => ({
		title: strings('app_settings.title'),
		headerTitleStyle: {
			fontSize: 20,
			...fontStyles.normal
		}
	});

	static propTypes = {
		/**
		/* navigation object required to push new views
		*/
		navigation: PropTypes.object
	};

	goTo = url => {
		this.props.navigation.pop();
		InteractionManager.runAfterInteractions(() => {
			this.props.navigation.navigate('BrowserView', {
				url
			});
		});
	};

	onPrivacyPolicy = () => {
		const url = 'https://metamask.io/privacy.html';
		this.goTo(url);
	};

	onTermsOfUse = () => {
		const url = 'https://metamask.io/terms.html';
		this.goTo(url);
	};

	onAttributions = () => {
		const url = 'https://metamask.io/attributions.html';
		this.goTo(url);
	};

	onSupportCenter = () => {
		const url = 'https://metamask.zendesk.com/hc/en-us';
		this.goTo(url);
	};

	onWebSite = () => {
		const url = 'https://metamask.io/';
		this.goTo(url);
	};

	onEmailUs = () => {
		this.props.navigation.pop();
		Linking.openURL('mailto:help@metamask.io?subject=Feedback');
	};

	render() {
		return (
			<SafeAreaView style={styles.wrapper} testID={'app-settings-screen'}>
				<ScrollView contentContainerStyle={styles.wrapperContent}>
					<View style={styles.logoWrapper}>
						<Image source={foxImage} style={styles.image} resizeMethod={'auto'} />
					</View>
					<Text style={styles.title}>{strings('app_information.links')}</Text>
					<View style={styles.links}>
						<TouchableOpacity onPress={this.onPrivacyPolicy}>
							<Text style={styles.link}>{strings('app_information.privacy_policy')}</Text>
						</TouchableOpacity>
						<TouchableOpacity onPress={this.onTermsOfUse}>
							<Text style={styles.link}>{strings('app_information.terms_of_use')}</Text>
						</TouchableOpacity>
						<TouchableOpacity onPress={this.onAttributions}>
							<Text style={styles.link}>{strings('app_information.attributions')}</Text>
						</TouchableOpacity>
					</View>
					<View style={styles.division} />
					<View style={styles.links}>
						<TouchableOpacity onPress={this.onSupportCenter}>
							<Text style={styles.link}>{strings('app_information.support_center')}</Text>
						</TouchableOpacity>
						<TouchableOpacity onPress={this.onWebSite}>
							<Text style={styles.link}>{strings('app_information.web_site')}</Text>
						</TouchableOpacity>
						<TouchableOpacity onPress={this.onEmailUs}>
							<Text style={styles.link}>{strings('app_information.email_us')}</Text>
						</TouchableOpacity>
					</View>
				</ScrollView>
			</SafeAreaView>
		);
	}
}
