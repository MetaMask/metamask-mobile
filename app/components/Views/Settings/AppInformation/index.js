/* eslint-disable dot-notation */
import React, { PureComponent } from 'react';
import {
  SafeAreaView,
  StyleSheet,
  Image,
  Text,
  InteractionManager,
  View,
  ScrollView,
  TouchableOpacity,
} from 'react-native';
import {
  getApplicationName,
  getVersion,
  getBuildNumber,
} from 'react-native-device-info';
import { fontStyles } from '../../../../styles/common';
import PropTypes from 'prop-types';
import { strings } from '../../../../../locales/i18n';
import { getNavigationOptionsTitle } from '../../../UI/Navbar';
import AppConstants from '../../../../core/AppConstants';
import { ThemeContext, mockTheme } from '../../../../util/theme';

const IS_QA = process.env['METAMASK_ENVIRONMENT'] === 'qa';

const createStyles = (colors) =>
  StyleSheet.create({
    wrapper: {
      backgroundColor: colors.background.default,
      flex: 1,
    },
    wrapperContent: {
      paddingLeft: 24,
      paddingRight: 24,
      paddingVertical: 24,
    },
    title: {
      fontSize: 18,
      textAlign: 'left',
      marginBottom: 20,
      ...fontStyles.normal,
      color: colors.text.default,
    },
    link: {
      fontSize: 18,
      textAlign: 'left',
      marginBottom: 20,
      ...fontStyles.normal,
      color: colors.primary.default,
    },
    division: {
      borderBottomColor: colors.border.muted,
      borderBottomWidth: 1,
      width: '30%',
      marginBottom: 20,
    },
    image: {
      width: 100,
      height: 100,
    },
    logoWrapper: {
      flex: 1,
      backgroundColor: colors.background.default,
      alignItems: 'center',
      justifyContent: 'center',
      top: 20,
      marginBottom: 40,
    },
    versionInfo: {
      marginTop: 20,
      fontSize: 18,
      textAlign: 'left',
      marginBottom: 20,
      color: colors.text.alternative,
      ...fontStyles.normal,
    },
    branchInfo: {
      fontSize: 18,
      textAlign: 'left',
      marginBottom: 20,
      color: colors.text.alternative,
      ...fontStyles.normal,
    },
  });

const foxImage = require('../../../../images/fox.png'); // eslint-disable-line import/no-commonjs

/**
 * View that contains app information
 */
export default class AppInformation extends PureComponent {
  static propTypes = {
    /**
    /* navigation object required to push new views
    */
    navigation: PropTypes.object,
  };

  state = {
    appInfo: '',
  };

  updateNavBar = () => {
    const { navigation } = this.props;
    const colors = this.context.colors || mockTheme.colors;
    navigation.setOptions(
      getNavigationOptionsTitle(
        strings('app_settings.info_title'),
        navigation,
        false,
        colors,
      ),
    );
  };

  componentDidMount = async () => {
    this.updateNavBar();
    const appName = await getApplicationName();
    const appVersion = await getVersion();
    const buildNumber = await getBuildNumber();
    this.setState({ appInfo: `${appName} v${appVersion} (${buildNumber})` });
  };

  componentDidUpdate = () => {
    this.updateNavBar();
  };

  goTo = (url, title) => {
    InteractionManager.runAfterInteractions(() => {
      this.props.navigation.navigate('Webview', {
        screen: 'SimpleWebview',
        params: {
          url,
          title,
        },
      });
    });
  };

  onPrivacyPolicy = () => {
    const url = AppConstants.URLS.PRIVACY_POLICY;
    this.goTo(url, strings('app_information.privacy_policy'));
  };

  onTermsOfUse = () => {
    const url = AppConstants.URLS.TERMS_AND_CONDITIONS;
    this.goTo(url, strings('app_information.terms_of_use'));
  };

  onAttributions = () => {
    const url =
      'https://raw.githubusercontent.com/MetaMask/metamask-mobile/main/attribution.txt';
    this.goTo(url, strings('app_information.attributions'));
  };

  onSupportCenter = () => {
    const url = 'https://metamask.zendesk.com/hc/en-us';
    this.goTo(url, strings('drawer.metamask_support'));
  };

  onWebSite = () => {
    const url = 'https://metamask.io/';
    this.goTo(url, 'metamask.io');
  };

  onContactUs = () => {
    const url = 'https://metamask.zendesk.com/hc/en-us';
    this.goTo(url, strings('drawer.metamask_support'));
  };

  render = () => {
    const colors = this.context.colors || mockTheme.colors;
    const styles = createStyles(colors);

    return (
      <SafeAreaView style={styles.wrapper} testID={'app-settings-screen'}>
        <ScrollView contentContainerStyle={styles.wrapperContent}>
          <View style={styles.logoWrapper}>
            <Image
              source={foxImage}
              style={styles.image}
              resizeMethod={'auto'}
            />
            <Text style={styles.versionInfo}>{this.state.appInfo}</Text>
            {IS_QA ? (
              <Text style={styles.branchInfo}>
                {`Branch: ${process.env['GIT_BRANCH']}`}
              </Text>
            ) : null}
          </View>
          <Text style={styles.title}>{strings('app_information.links')}</Text>
          <View style={styles.links}>
            <TouchableOpacity onPress={this.onPrivacyPolicy}>
              <Text style={styles.link}>
                {strings('app_information.privacy_policy')}
              </Text>
            </TouchableOpacity>
            <TouchableOpacity onPress={this.onTermsOfUse}>
              <Text style={styles.link}>
                {strings('app_information.terms_of_use')}
              </Text>
            </TouchableOpacity>
            <TouchableOpacity onPress={this.onAttributions}>
              <Text style={styles.link}>
                {strings('app_information.attributions')}
              </Text>
            </TouchableOpacity>
          </View>
          <View style={styles.division} />
          <View style={styles.links}>
            <TouchableOpacity onPress={this.onSupportCenter}>
              <Text style={styles.link}>
                {strings('app_information.support_center')}
              </Text>
            </TouchableOpacity>
            <TouchableOpacity onPress={this.onWebSite}>
              <Text style={styles.link}>
                {strings('app_information.web_site')}
              </Text>
            </TouchableOpacity>
            <TouchableOpacity onPress={this.onContactUs}>
              <Text style={styles.link}>
                {strings('app_information.contact_us')}
              </Text>
            </TouchableOpacity>
          </View>
        </ScrollView>
      </SafeAreaView>
    );
  };
}

AppInformation.contextType = ThemeContext;
