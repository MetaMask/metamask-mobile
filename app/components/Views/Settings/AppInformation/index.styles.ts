import { StyleSheet } from 'react-native';
import { fontStyles } from '../../../../styles/common';

export const createStyles = (colors: {
  background: { default: string };
  text: { default: string; alternative: string };
  primary: { default: string };
  border: { muted: string };
}) =>
  StyleSheet.create({
    wrapper: {
      backgroundColor: colors.background.default,
      flex: 1,
    },
    wrapperContent: {
      paddingLeft: 24,
      paddingRight: 24,
      paddingVertical: 24,
    },
    title: {
      fontSize: 18,
      textAlign: 'left',
      marginBottom: 20,
      ...fontStyles.normal,
      color: colors.text.default,
    },
    link: {
      fontSize: 18,
      textAlign: 'left',
      marginBottom: 20,
      ...fontStyles.normal,
      color: colors.primary.default,
    },
    division: {
      borderBottomColor: colors.border.muted,
      borderBottomWidth: 1,
      width: '30%',
      marginBottom: 20,
    },
    image: {
      width: 100,
      height: 100,
    },
    logoWrapper: {
      flex: 1,
      backgroundColor: colors.background.default,
      alignItems: 'center',
      justifyContent: 'center',
      top: 20,
      marginBottom: 40,
    },
    versionInfo: {
      marginTop: 20,
      fontSize: 18,
      textAlign: 'left',
      marginBottom: 20,
      color: colors.text.alternative,
      ...fontStyles.normal,
    },
    branchInfo: {
      fontSize: 18,
      textAlign: 'left',
      marginBottom: 20,
      color: colors.text.alternative,
      ...fontStyles.normal,
    },
    links: {
      // Container for links - no specific styling needed
    },
  });
