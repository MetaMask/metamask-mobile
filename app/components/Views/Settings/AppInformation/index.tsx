/* eslint-disable dot-notation */
import React, { useState, useEffect, useCallback, useContext } from 'react';
import {
  SafeAreaView,
  Image,
  Text,
  View,
  ScrollView,
  TouchableOpacity,
} from 'react-native';
import {
  getApplicationName,
  getVersion,
  getBuildNumber,
} from 'react-native-device-info';
import { strings } from '../../../../../locales/i18n';
import { getNavigationOptionsTitle } from '../../../UI/Navbar';
import AppConstants from '../../../../core/AppConstants';
import { ThemeContext, mockTheme } from '../../../../util/theme';
import { AboutMetaMaskSelectorsIDs } from '../../../../../e2e/selectors/Settings/AboutMetaMask.selectors';
import { useSupportConsent } from '../../../hooks/useSupportConsent';
import { isQa } from '../../../../util/test/utils';
import { createStyles } from './index.styles';

import foxImage from '../../../../images/branding/fox.png';

interface AppInformationProps {
  navigation: {
    navigate: (
      screen: string,
      params?: { screen: string; params: { url: string; title: string } },
    ) => void;
    setOptions: (options: { headerTitleAlign?: string }) => void;
  };
}

/**
 * View that contains app information
 */
const AppInformation = ({ navigation }: AppInformationProps) => {
  const colors = useContext(ThemeContext)?.colors || mockTheme.colors;
  const styles = createStyles(colors);

  const [appInfo, setAppInfo] = useState('');
  const [appVersion, setAppVersion] = useState('');

  // Simple navigation function for all links
  // without over-engineering with interaction manager
  // that not providing any benefit here as we just need to navigate to the webview
  const goTo = useCallback(
    (url: string, title: string) => {
      navigation.navigate('Webview', {
        screen: 'SimpleWebview',
        params: {
          url,
          title,
        },
      });
    },
    [navigation],
  );

  const { openSupportWebPage } = useSupportConsent(
    goTo,
    strings('drawer.metamask_support'),
  );

  const updateNavBar = useCallback(() => {
    navigation.setOptions(
      getNavigationOptionsTitle(
        strings('app_settings.info_title'),
        navigation,
        false,
        colors,
      ),
    );
  }, [navigation, colors]);

  useEffect(() => {
    updateNavBar();
  }, [updateNavBar]);

  useEffect(() => {
    const loadAppInfo = async () => {
      const appName = await getApplicationName();
      const version = await getVersion();
      const buildNumber = await getBuildNumber();
      setAppInfo(`${appName} v${version} (${buildNumber})`);
      setAppVersion(version);
    };

    loadAppInfo();
  }, []);

  const onPrivacyPolicy = useCallback(() => {
    const url = AppConstants.URLS.PRIVACY_POLICY;
    goTo(url, strings('app_information.privacy_policy'));
  }, [goTo]);

  const onTermsOfUse = useCallback(() => {
    const url = AppConstants.URLS.TERMS_AND_CONDITIONS;
    goTo(url, strings('app_information.terms_of_use'));
  }, [goTo]);

  const onAttributions = useCallback(() => {
    const url = `https://raw.githubusercontent.com/MetaMask/metamask-mobile/v${appVersion}/attribution.txt`;
    goTo(url, strings('app_information.attributions'));
  }, [goTo, appVersion]);

  // current design is to open the support center in the same way as contact us link,
  // so it also calls openSupportWebPage and triggers the consent flow
  // TODO: consider deleting this redundant link
  const onSupportCenter = useCallback(() => {
    openSupportWebPage();
  }, [openSupportWebPage]);

  const onWebSite = useCallback(() => {
    const url = 'https://metamask.io/';
    goTo(url, 'metamask.io');
  }, [goTo]);

  const onContactUs = useCallback(() => {
    openSupportWebPage();
  }, [openSupportWebPage]);

  return (
    <SafeAreaView
      style={styles.wrapper}
      testID={AboutMetaMaskSelectorsIDs.CONTAINER}
    >
      <ScrollView contentContainerStyle={styles.wrapperContent}>
        <View style={styles.logoWrapper}>
          <Image source={foxImage} style={styles.image} resizeMethod={'auto'} />
          <Text style={styles.versionInfo}>{appInfo}</Text>
          {isQa ? (
            <Text style={styles.branchInfo}>
              {`Branch: ${process.env['GIT_BRANCH']}`}
            </Text>
          ) : null}
        </View>
        <Text style={styles.title}>{strings('app_information.links')}</Text>
        <View style={styles.links}>
          <TouchableOpacity onPress={onPrivacyPolicy}>
            <Text style={styles.link}>
              {strings('app_information.privacy_policy')}
            </Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={onTermsOfUse}>
            <Text style={styles.link}>
              {strings('app_information.terms_of_use')}
            </Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={onAttributions}>
            <Text style={styles.link}>
              {strings('app_information.attributions')}
            </Text>
          </TouchableOpacity>
        </View>
        <View style={styles.division} />
        <View style={styles.links}>
          <TouchableOpacity onPress={onSupportCenter}>
            <Text style={styles.link}>
              {strings('app_information.support_center')}
            </Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={onWebSite}>
            <Text style={styles.link}>
              {strings('app_information.web_site')}
            </Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={onContactUs}>
            <Text style={styles.link}>
              {strings('app_information.contact_us')}
            </Text>
          </TouchableOpacity>
        </View>
      </ScrollView>
    </SafeAreaView>
  );
};

export default AppInformation;
