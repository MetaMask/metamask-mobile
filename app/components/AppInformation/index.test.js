import React from 'react';
import { shallow } from 'enzyme';
import AppInformation from './';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

const mockStore = configureMockStore();
const store = mockStore({});

describe('AppInformation', () => {
	it('should render correctly', () => {
		const wrapper = shallow(
			<Provider store={store}>
				<AppInformation navigation={{ state: { params: {} } }} />
			</Provider>
		);
		expect(wrapper).toMatchSnapshot();
	});
});
