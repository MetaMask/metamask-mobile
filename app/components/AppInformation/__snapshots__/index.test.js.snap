// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInformation should render correctly 1`] = `
<AppInformation
  navigation={
    Object {
      "state": Object {
        "params": Object {},
      },
    }
  }
/>
`;
